pkg/security/utils/path.go

```diff
@@ -0,0 +1,115 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux
+
+package utils
+
+// PathPatternBuilderOpts PathPatternBuilder options
+type PathPatternBuilderOpts struct {
+	WildcardLimit      int // max number of wildcard in the pattern
+	PrefixNodeRequired int // number of prefix nodes required
+	SuffixNodeRequired int // number of suffix nodes required
+	NodeSizeLimit      int // min size required to substitute with a wildcard
+}
+
+// PathPatternBuilder pattern builder for files
+func PathPatternBuilder(pattern string, path string, opts PathPatternBuilderOpts) (bool, string) {
+	lenMax := len(pattern)
+	if l := len(path); l > lenMax {
+		lenMax = l
+	}
+
+	var (
+		i, j                                 = 0, 0
+		result                               = make([]byte, lenMax)
+		offsetPattern, offsetPath, size      = 0, 0, 0
+		wildcardCount, nodeCount, suffixNode = 0, 0, 0
+		wildcard, inPattern, inPath          bool
+
+		computeNode = func() bool {
+			if wildcard {
+				wildcardCount++
+				if wildcardCount > opts.WildcardLimit {
+					return false
+				}
+				if nodeCount < opts.PrefixNodeRequired {
+					return false
+				}
+				if opts.NodeSizeLimit != 0 && j-offsetPath < opts.NodeSizeLimit {
+					return false
+				}
+
+				result[size], result[size+1] = '/', '*'
+				size += 2
+
+				offsetPattern, suffixNode = i, 0
+			} else {
+				copy(result[size:], pattern[offsetPattern:i])
+				size += i - offsetPattern
+				offsetPattern = i
+				suffixNode++
+			}
+			offsetPath = j
+
+			if i > 0 {
+				nodeCount++
+			}
+			return true
+		}
+	)
+
+	for i != len(pattern) && j != len(path) {
+		pn, ph := pattern[i], path[j]
+		if pn == '/' && !inPattern {
```

Comment:
If I'm not mistaken, the first 2 `if` statements are used only for the first `/` match (which is pretty much bound to happen at the beginning of the path) ... which means that you should be able to get rid of it entirely to avoid the branching and speed everything up ! 😅 See the diff below !

```
diff --git a/pkg/security/utils/path.go b/pkg/security/utils/path.go
index 5d332da443..c63652fb6b 100644
--- a/pkg/security/utils/path.go
+++ b/pkg/security/utils/path.go
@@ -27,7 +27,7 @@ func PathPatternBuilder(pattern string, path string, opts PathPatternBuilderOpts
 		result                               = make([]byte, lenMax)
 		offsetPattern, offsetPath, size      = 0, 0, 0
 		wildcardCount, nodeCount, suffixNode = 0, 0, 0
-		wildcard, inPattern, inPath          bool
+		wildcard                             bool
 
 		computeNode = func() bool {
 			if wildcard {
@@ -45,13 +45,13 @@ func PathPatternBuilder(pattern string, path string, opts PathPatternBuilderOpts
 				result[size], result[size+1] = '/', '*'
 				size += 2
 
-				offsetPattern, suffixNode = i, 0
+				suffixNode = 0
 			} else {
 				copy(result[size:], pattern[offsetPattern:i])
 				size += i - offsetPattern
-				offsetPattern = i
 				suffixNode++
 			}
+			offsetPattern = i
 			offsetPath = j
 
 			if i > 0 {
@@ -63,14 +63,15 @@ func PathPatternBuilder(pattern string, path string, opts PathPatternBuilderOpts
 
 	for i != len(pattern) && j != len(path) {
 		pn, ph := pattern[i], path[j]
-		if pn == '/' && !inPattern {
-			offsetPattern = i
-			inPattern = true
-		}
+		if pn == '/' && ph == '/' {
+			if !computeNode() {
+				return false, ""
+			}
+			wildcard = false
 
-		if ph == '/' && !inPath {
-			offsetPath = j
-			inPath = true
+			i++
+			j++
+			continue
 		}
 
 		if pn != ph {
@@ -90,17 +91,6 @@ func PathPatternBuilder(pattern string, path string, opts PathPatternBuilderOpts
 			}
 			break
 		}
-
-		pn, ph = pattern[i], path[j]
-		if pn == '/' && ph == '/' {
-			if !computeNode() {
-				return false, ""
-			}
-			wildcard, inPattern = false, false
-
-			i++
-			j++
-		}
 	}
 
 	if !computeNode() {
```

With this patch, I got a ~30% execution time improvement 🎉 

Before:
```
/tmp/GoLand/___BenchmarkPathPatternBuilder_in_github_com_DataDog_datadog_agent_pkg_security_utils.test -test.v -test.paniconexit0 -test.bench ^\QBenchmarkPathPatternBuilder\E$ -test.run ^$
goos: linux
goarch: amd64
pkg: github.com/DataDog/datadog-agent/pkg/security/utils
cpu: 11th Gen Intel(R) Core(TM) i9-11950H @ 2.60GHz
BenchmarkPathPatternBuilder
BenchmarkPathPatternBuilder/pattern
BenchmarkPathPatternBuilder/pattern-16          11680924               122.6 ns/op
BenchmarkPathPatternBuilder/standard
BenchmarkPathPatternBuilder/standard-16         1000000000               0.2655 ns/op
PASS
```

After:
```
/tmp/GoLand/___BenchmarkPathPatternBuilder_in_github_com_DataDog_datadog_agent_pkg_security_utils.test -test.v -test.paniconexit0 -test.bench ^\QBenchmarkPathPatternBuilder\E$ -test.run ^$
goos: linux
goarch: amd64
pkg: github.com/DataDog/datadog-agent/pkg/security/utils
cpu: 11th Gen Intel(R) Core(TM) i9-11950H @ 2.60GHz
BenchmarkPathPatternBuilder
BenchmarkPathPatternBuilder/pattern
BenchmarkPathPatternBuilder/pattern-16          13729456                87.93 ns/op
BenchmarkPathPatternBuilder/standard
BenchmarkPathPatternBuilder/standard-16         1000000000               0.2102 ns/op
PASS
```

Reply:
Nice improvement, thanks